### PR TITLE
feat: add runtime validation for API request parameters

### LINKED ISSUE #15: Add runtime validation for API request parameters
## Issue

API routes use TypeScript type assertions to cast request parameters without runtime validation:

```typescript
// src/server/routes/accounts/delete-account.ts
const account_id = req.query.id as string;

// src/server/routes/accounts/post-transaction.ts
await updateTransactions(user, [req.body]);  // body passed directly
```

### Problems

1. **Missing parameters**: `undefined` cast as `string` passes type checking
2. **Array params**: Express parses `?id=a&id=b` as array, not string
3. **Type coercion**: Numbers, booleans passed as strings
4. **Unvalidated body**: Request body properties not checked before DB operations

### Suggested Approach

Add lightweight validation using Zod or manual checks:

```typescript
import { z } from 'zod';

const deleteAccountSchema = z.object({
  id: z.string().uuid()
});

// In route handler:
const result = deleteAccountSchema.safeParse(req.query);
if (!result.success) {
  return { status: 'failed', message: 'Invalid account ID' };
}
const { id: account_id } = result.data;
```

Or simpler manual approach:
```typescript
const account_id = req.query.id;
if (typeof account_id !== 'string' || !account_id) {
  return { status: 'failed', message: 'Missing account ID' };
}
```

### Impact

Low severity for internal use, but good practice for robustness and clearer error messages.

### PR DESCRIPTION
**Pull request by moltboie**
## Summary
Addresses #15 - API routes now validate request parameters at runtime instead of using unsafe type assertions.

### New validation utilities (`src/server/lib/validation.ts`)
- `requireQueryString`: validates required string query params
- `optionalQueryString`: validates optional string query params  
- `requireBodyObject`: validates request body is an object
- `requireStringField/requireNumberField`: validates object fields
- `validationError`: helper for consistent error responses

### Updated routes
- `delete-account.ts`: validates 'id' parameter
- `delete-chart.ts`: replaces manual check with validation helper
- `delete-snapshot.ts`: adds missing validation (was casting undefined as string)
- `get-transactions.ts`: validates optional date/account params
- `post-transaction.ts`: validates body and transaction_id field

### Tests
Comprehensive test suite for all validation helpers with 19 test cases covering:
- Valid inputs
- Missing parameters
- Array parameters (Express edge case)
- Empty strings
- Type mismatches
- Edge cases (NaN, Infinity)

## Impact
This prevents:
- Missing parameters being cast as strings silently
- Array parameters (from `?id=a&id=b`) being mishandled
- Invalid body types being passed to DB operations

Lightweight approach with no external dependencies - just manual type checks.

Closes #15

### CODE REVIEW

File: src/server/lib/validation.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+ * Extract an optional string parameter from request query.
+ * Returns undefined if not present.
+ */
+export function optionalQueryString(
+  req: Request,
+  param: string
+): ValidationResult<string | undefined> {
+  const value = req.query[param];
+
+  if (value === undefined || value === null) {
```
Comment (hoiekim) [inline src/server/lib/validation.ts]: Same comment as above. Use the type checking utils that we already wrote in the common module. Apply this to all type checking conditions in this PR

**Comment by moltboie** [comment]
Addressed both questions in c364921:

**1. Why don't we add validation to all routes?**
Fixed - added validation to all 23 remaining routes that accept parameters. Every route now uses the validation helpers consistently.

**2. Did you test end-to-end behavior?**
- TypeScript compiles cleanly
- All 24 unit tests pass
- Manual testing recommended before merge:
  - Create/update/delete operations
  - Invalid parameter handling (missing required, wrong types)
  - Empty request bodies

The validation now covers:
- All DELETE routes (budget, category, section, item, split-transaction, etc.)
- All POST routes (budget, category, section, chart, login, transactions, etc.)
- All GET routes with parameters (institution, snapshots, link-token, etc.)
- Webhook endpoint (plaid-hook)

File: src/server/lib/validation.ts (hunk 1 above)
Comment (moltboie) [reply src/server/lib/validation.ts]: Fixed in the same commit (b003090) - all type checks now use common module utilities.

File: src/server/lib/validation.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+  // Express can parse ?id=a&id=b as an array
+  if (Array.isArray(value)) {
+    return {
+      success: false,
+      error: `Parameter ${param} must be a single value, not an array`,
+    };
+  }
+
+  if (typeof value !== "string") {
```
[the remark this answers is not in the document]
Comment (moltboie) [reply src/server/lib/validation.ts]: Done - refactored to use `isString`, `isNumber`, `isArray`, etc. from common module in commit b003090.
